docs/fundamentals/gateway/custom-gateway.md

```diff
@@ -0,0 +1,326 @@
+```python
+from jina import Gateway
+
+
+class MyGateway(Gateway):
+    def __init__(self, foo: str, **kwargs):
+        super().__init__(**kwargs)
+        self.foo = foo
+```
+
+* Implement `async def setup_server():`. This should set up a server runnable on an asyncio loop (and other resources 
+needed for setting up the server). For instance:
+```python
+from jina import Gateway
+from fastapi import FastAPI
+from uvicorn import Server, Config
+
+
+class MyGateway(Gateway):
+    async def setup_server(self):
+        app = FastAPI(title='My Custom Gateway')
+
+        @app.get(path='/endpoint')
+        def custom_endpoint():
+            return {'message': 'custom-gateway'}
+
+        self.server = Server(Config(app, host=self.host, port=self.port))
+```
+* Implement `async def run_server():`. This should run the server and await for it while serving:
+```python
+from jina import Gateway
+
+
+class MyGateway(Gateway):
+    ...
+
+    async def run_server(self):
+        await self.server.serve()
+```
+* Implement `async def shutdown():`. This should run the server and await for it while serving:
+```python
+from jina import Gateway
+
+
+class MyGateway(Gateway):
+    ...
+
+    async def shutdown(self):
+        self.server.should_exit = True
+        await self.server.shutdown()
+```
+
+As an example, you can refer to {class}`~jina.serve.runtimes.gateway.grpc.GRPCGateway` and 
+{class}`~jina.serve.runtimes.gateway.websocket.WebSocketGateway`.
+
+### Using {class}`~jina.serve.runtimes.gateway.http.fastapi.FastAPIBaseGateway`:
+{class}`~jina.serve.runtimes.gateway.http.fastapi.FastAPIBaseGateway` offers a simpler API to implement custom 
+gateways but is restricted to FastAPI apps.
+
+In order to implement a custom gateway using {class}`~jina.jina.serve.runtimes.gateway.http.fastapi.FastAPIBaseGateway`, 
+simply implement the {meth}`~jina.jina.serve.runtimes.gateway.http.fastapi.FastAPIBaseGateway.app` property:
+
+```python
+from jina.serve.runtimes.gateway.http.fastapi import FastAPIBaseGateway
+
+
+class DummyFastAPIGateway(FastAPIBaseGateway):
+    @property
+    def app(self):
+        from fastapi import FastAPI
+
+        app = FastAPI(title='Custom FastAPI Gateway')
+
+        @app.get(path='/endpoint')
+        def custom_endpoint():
+            return {'message': 'custom-fastapi-gateway'}
+
+        return app
+```
+
+As an example, you can refer to {class}`~jina.serve.runtimes.gateway.http.HTTPGateway`.
+
+{class}`~jina.serve.runtimes.gateway.http.fastapi.FastAPIBaseGateway` is a subclass of {class}`~jina.Gateway` and therefore 
+shares its attributes.
+
+## Gateway arguments
+### Runtime attributes
+Jina injects some attributes into the Gateway classes. You can use them to set up your custom gateway:
+* name: gateway pod name.
+* logger: Jina logger object.
+* tracing: whether runtime tracing is enabled or not.
+* tracer_provider: OpenTelemetry `TraceProvider` object.
```

Comment:
I would highlight the ones that are more usable and separate from those that are more advanced

Reply:
Okay I'm hiding tracing and tracing provider